### PR TITLE
[stdlib] Implement oracle data protocol

### PR DESCRIPTION
**Pull request by jolestar**
1. 只考虑了数据格式的定义和更新
2. 如果要实现去中心化 Oracle 协议，可以把 UpdatePriceCapability 托管到另外的Module 中，然后设计更新协议。
3. 只实现了简单的一种聚合方法，未考虑历史数据的获取和更新。

### CODE REVIEW

File: vm/stdlib/modules/Oracle.move [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    struct UpdatePriceCapability<DataT: copy+store+drop> has store, key {
+        account: address,
+    }
+
+    public fun register_price_oracle<DataT: copy+store+drop>(signer: &signer, precision: u8, description: vector<u8>){
+        //TODO implement a global register by contact account.
+        CoreAddresses::assert_genesis_address(signer);
+        move_to(signer, PriceOracleInfo<DataT> {
```
Comment (wh-antares) [inline vm/stdlib/modules/Oracle.move:49]: should we define a constant to limit the range of precision?  like MAX_DECIMALS

File: vm/stdlib/modules/Oracle.move [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    public fun read_price<DataT:copy+store+drop>(addr: address): u128 acquires PriceFeed{
+        let price_feed = borrow_global<PriceFeed<DataT>>(addr);
+        *&price_feed.value.price
+    }
+
+    public fun read_price_data<DataT:copy+store+drop>(addr: address): PriceData acquires PriceFeed{
+        let price_feed = borrow_global<PriceFeed<DataT>>(addr);
+        *&price_feed.value
```
Comment (wh-antares) [inline vm/stdlib/modules/Oracle.move:105]: Maybe  this return value should be a struct containing the precision and the number after pow.
Like `  struct Num { value: u64, dec: u64 }` . .
So is there another question, is there a FixedPoint64 similar to FixedPoint32 in stdlib?

Comment (jolestar) [reply vm/stdlib/modules/Oracle.move:105]: PriceData.price 保存的是原始数据，precision 是用来展示分数的时候才需要的

File: vm/stdlib/modules/Oracle.move (hunk 1 above)
Comment (jolestar) [reply vm/stdlib/modules/Oracle.move:49]: 不好确定 MAX_DECIMALS 的值，理论上 Math::pow(10, precision) < u128::max_value() 就可以吧

**Comment by jolestar** [comment]
> Can it be generalized to store any data, instead of only num value.
> I imagine it to be a generalized data oracle, which can also store other values like hash root, structured data.
> 
> Then on top of it, we can do another abstraction, like price oracle to do price related oracle, merkle root oracle to do block changing oracle。

我开始也是这样想的，但那样设计通用的方法比较困难。我尝试另外一个思路。

**Comment by jolestar** [comment]
> Can it be generalized to store any data, instead of only num value.
> I imagine it to be a generalized data oracle, which can also store other values like hash root, structured data.
> 
> Then on top of it, we can do another abstraction, like price oracle to do price related oracle, merkle root oracle to do block changing oracle。

#2732 中提供了另外一种实现

**Pull request closed without merging**